src/util/indexing.jl

```diff
@@ -0,0 +1,163 @@
+module Indexing
+
+export
+    AbstractIndex, SingleIndex, names!, rename!, rename
+
+
+# Indexing datastructures
+
+# Abstract Types
+abstract AbstractIndex
+
+
+# Type Aliases and unions
+typealias TableRange UnitRange{Int}
+rangeOrVect = Union{TableRange, Vector{Int}}
+
+
+# Concrete Types
+type SingleIndex <: AbstractIndex
+    lookup::Dict{Symbol, Int}
+    names::Vector{Symbol}
+end
+
+function SingleIndex(names::Vector{Symbol})
+    u = make_unique(names)
+    lookup = Dict{Symbol, Int}(zip(u, 1:length(u)))
+    SingleIndex(lookup, u)
+end
+SingleIndex{S <: AbstractString}(names::Vector{S}) = SingleIndex(convert(Vector{Symbol}, names))
+SingleIndex() = SingleIndex(Dict{Symbol, Int}(), Symbol[])
+
+
+type GroupIndex{T <: rangeOrVect} <: AbstractIndex
```

Comment:
I didn't know UnitRange was an abstract vector, although it does now make sense when I think LightGraphs makes use of UnitRange for it's vertex storage! Thanks @bicycle1885!